### DISCUSSION TITLE
V2 Protected Route for Next.js issue

### DISCUSSION THREAD

**Discussion by brentrobbins** [nhost/nhost#97] [Q&A]
Hello. When I login (using the code supplied in the docs) for V2 in my Next.js project, it does route/redirect me to my app even though it's responding with a successful response along with my session and all of the details. So if I refresh the page a couple times and try to login it eventually works sending me to the app. I am using the protected route code supplied in the docs. Here's a [screencast](https://watch.screencastify.com/v/bAi89gRrFhsSB5iiNQ9S) to help demonstrate what I am experiencing locally.

**Comment by st3phan** [comment]
Sounds like a race condition. Can you share some code? Your protected route component and your sign in component maybe?
It's difficult to tell from the video only what the problem might be.
You can also find me on discord if you prefer a chat: st3phan.

**Comment by elitan** [comment]
Yea some code would be great to help debug. 

Do you use the code from here: https://docs.nhost.io/reference/supporting-libraries/react-auth#next-js ?

Let's try to keep the conversation here at GitHub to make the conversation searchable so it can help future developers with similar issues.

**Comment by brentrobbins** [comment]
Here's my code as previously mentioned it's the same as the one from the docs. I just added my own loader.

```
import { useRouter } from 'next/router'
import { useNhostAuth } from '@nhost/react-auth'
import { Loader } from '@/components/shared/loader'

export function PrivateRoute(Comp) {
  return function AuthProtected(props) {
    const router = useRouter()
    const { isLoading, isAuthenticated } = useNhostAuth()
    if (isLoading) {
      return <Loader />
    }

    if (!isAuthenticated) {
      router.push('/login')
      return null
    }

    return <Comp {...props} />
  };
}
```

**Reply by st3phan** [reply]
And what does your login component look like?

**Reply by brentrobbins** [reply]
Sure thing, here it is:

The login function:
```
  async function handleSubmit(e: { preventDefault: () => void; }) {
    e.preventDefault();
    setIsLoading(true);
    let response
    try {
      response = await nhost.auth.signIn({ email, password });
      if (response.status >= 400) {
        const errorMessage = response?.error || SiteConst.LOGIN_ERROR_MESSAGE;
        notify("error", errorMessage);
        setIsLoading(false);
        return;
      }
    } catch (error) {
      const errorMessage = error?.response?.data?.message || SiteConst.LOGIN_ERROR_MESSAGE;
      notify("error", errorMessage);
      setIsLoading(false);
      return;
    }
      notify("success", SiteConst.LOGIN_SUCCESS_MESSAGE);
      setIsLoading(false);
      Router.push('/app')
  }
```

**Reply by st3phan** [reply]
Looks ok to me but are you sure `Router.push('/app')` is not firing too soon?
I would put the `Router.push('/app')` in the try block though so it's more obvious what all happy flow code is.

Also I'm wondering what you routes logic looks like because what I find interesting is that after you click the submit button and the call is successful it doesn't redirect you... Or is the redirect happening but your protected is redirecting you back to the login page?

**Reply by brentrobbins** [reply]
@st3phan I moved the route.push into the try (which I actually had before and it didn't help before). This time it again didn't change anything and produces the same results. I recorded a [screencast](https://watch.screencastify.com/v/NKC9DQ8R24dc4jIt0uXG) to show you what I am experiencing. 

I am not sure what you mean by my route logic not working because I think it is working because it's routing to the app page that's wrapped with the private route component and I added a console log to make sure and for some reason it's showing `isAuthenticated: false` and that's why it's kicking me back to the login route. What's odd is I'll try again with the same credentials and then it'll actually work. You'll see that happen in the screencast. 

Here's my update login logic:
```
  async function handleSubmit(e: { preventDefault: () => void; }) {
    e.preventDefault();
    setIsLoading(true);
    let response
    try {
      response = await nhost.auth.signIn({ email, password });
      if (response.error) {
        const errorMessage = SiteConst.LOGIN_ERROR_MESSAGE;
        notify("error", errorMessage);
        setIsLoading(false);
        return;
      } else {
        notify("success", SiteConst.LOGIN_SUCCESS_MESSAGE);
        setIsLoading(false);
        Router.push('/app')
      }

    } catch (error) {
      const errorMessage = SiteConst.LOGIN_ERROR_MESSAGE;
      notify("error", errorMessage);
      setIsLoading(false);
      return;
    }
  }
```

**Reply by st3phan** [reply]
Hmm I don't see anything strange is your code. Maybe one thing that I notice is that the error handling in the `try` seems redundant. Isn't that handled by the `catch`?

Would it be possible to create public repo for us to checkout? Makes it easier to debug.

**Reply by st3phan** [reply]
Hmm I just found out the v2 auth methods are not throwing errors anymore compared to v1.
So my comment on the error handling in the try catch block is wrong.

**Reply by brentrobbins** [reply]
OK, well I updated it in my previous message. If my code isn't correct, could you please let me know what it should be?

By the way, I am working on a public repo I can share.

**Reply by brentrobbins** [reply]
Here's a public repo with a basic Nhost v2 with Next.js setup. I thinking the issue is when using Preact. I tried it with standard React and it worked fine, but I am not sure why it wouldn't work with Preact too. Also that's just my guess. Thanks for your help!

https://github.com/brentrobbins/Nhost-V2-Demo

**Reply by st3phan** [reply]
It looks like `onAuthStateChanged` in `NhostAuthProvider` from the `@nhost/react-auth` package is not working with preact. I've never used preact so not sure if what to do here. Looks like something the nhost guys might need to take a look at.

**Reply by brentrobbins** [reply]
OK, @st3phan. Thanks for your help narrowing things down.

☝🏻  @elitan or @nunopato, I would really appreciate the help 🙏🏻

**Reply by elitan** [reply]
@brentrobbins Thanks for putting together this repo. It was really easy to re-produce and I am indeed able to reproduce so it seems to be a bug on our end. We're investigating. Sorry for it taking some time.

**Comment by sangdth** [comment]
Update 1:
I remember it happens more often if in the development process my app crashes (the red error screen that Nextjs show error messages), from that point, usually it likely happens.


Sometimes I have this issue too. I realize if I refresh it with force (Cmd + Shift + R) it might get back to work. Normal refresh does not help. I will update with better info later, as it happens quite randomly.

**Comment by plmercereau** [comment]
Hello,
We created a new `@nhost/nextjs` package, please have a look at its documentation:
https://docs.nhost.io/reference/nextjs
